### PR TITLE
Fixed einsum compatibility/performance issues (#46398)

### LINKED ISSUE #32591: `einsum` is ~20X slower than manually multiplying and summing
## 🐛 Bug
A manual multiplication and summation `(a * b).sum(dim = (-3, -2, -1))` is about 20X faster than the equivalent `einsum`.

## To Reproduce

```Python3
a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device = "cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device = "cuda")

# Warmup
for i in range(100): output1 = (a * b).sum(dim = (-3, -2, -1))

# Method 1
with torch.autograd.profiler.profile(use_cuda = True) as prof:
  output1 = (a * b).sum(dim = (-3, -2, -1))
print(prof.key_averages().table(sort_by="cuda_time_total"))

# Warmup
for i in range(100):
  (a2, b2) = torch.broadcast_tensors(a, b)
  output2 = torch.einsum("...ijk, ...ijk -> ...", a2, b2)

# Method 2
with torch.autograd.profiler.profile(use_cuda = True) as prof:
  (a2, b2) = torch.broadcast_tensors(a, b)
  output2 = torch.einsum("...ijk, ...ijk -> ...", a2, b2)
print(prof.key_averages().table(sort_by="cuda_time_total"))
```
```
-------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
-------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
sum      28.93%           43.848us         28.93%           43.848us         43.848us         65.81%           8.978ms          8.978ms          1                
mul      71.07%           107.713us        71.07%           107.713us        107.713us        34.19%           4.665ms          4.665ms          1                
-------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 151.561us
CUDA time total: 13.644ms

---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                   Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
einsum                 4.05%            59.449us         97.88%           1.435ms          1.435ms          48.62%           123.798ms        123.798ms        1                
bmm                    85.15%           1.248ms          85.15%           1.248ms          1.248ms          45.83%           116.694ms        116.694ms        1                
reshape                1.35%            19.770us         6.36%            93.237us         46.619us         2.76%            7.031ms          3.516ms          2                
clone                  3.96%            58.021us         3.96%            58.021us         29.011us         2.75%            7.012ms          3.506ms          2                
broadcast_tensors      0.79%            11.624us         2.12%            31.095us         31.095us         0.01%            31.008us         31.008us         1                
expand                 1.33%            19.471us         1.33%            19.471us         9.736us          0.01%            20.480us         10.240us         2                
permute                1.54%            22.535us         1.54%            22.535us         4.507us          0.01%            19.327us         3.865us          5                
view                   0.78%            11.442us         0.78%            11.442us         5.721us          0.00%            4.094us          2.047us          2                
_unsafe_view           1.05%            15.446us         1.05%            15.446us         7.723us          0.00%            3.968us          1.984us          2                
---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 1.466ms
CUDA time total: 254.614ms
```

## Expected behavior

`einsum` to be no slower than manual. I actually expected / hoped for it to be faster as it avoids creating the large intermediary result of the multiplication.

## Environment

The environment is Google Colab with a fresh GPU instance on which `!pip install -q torch==1.4.0 torchvision==0.5.0` has been run as Colab is still on `1.3.1` by default.

```
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.17.5
[pip3] torch==1.4.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.5.0
[conda] Could not collect
```

cc @ngimel @vincentqb @vishwakftw @jianyuh @nikitaved @pearu @VitalyFedyunin @mruberry

### LINKED ISSUE #41467: einsum running 100x slower than expected when controlling for number of ops
## 🐛 Bug

For two simple einsum cases, Case 2 is running ~30% slower than Case 1, despite needing to perform 101x fewer ops.

## To Reproduce

**Case 1**
```
a = torch.randn(10000, 100, 101).cuda()
b = torch.randn(10000, 101, 3).cuda()

%%time
for _ in range(10000):
    torch.einsum("bij, bjf->bif", a, b)
    torch.cuda.synchronize()
```
`CPU times: user 26.7 s, sys: 18.8 s, total: 45.5 s
Wall time: 45.5 s`

**Case 2**
```
c = torch.randn(10000, 100, 1).cuda()
d = torch.randn(10000, 100, 1, 3).cuda()

%%time
for _ in range(1000):
    torch.einsum("bic, bicf->bif", c, d)
    torch.cuda.synchronize()
```
`CPU times: user 37.7 s, sys: 23.7 s, total: 1min 1s
Wall time: 1min 1s`

## Expected behavior

I believe the number of ops needed for each case should be approximately:
**Case 1:** 10,000 x 100 x 101 x 3 ( x 2)
**Case 2:** 10,000 x 100 x 1 x 3 (x 2) 
(The factor of 2 is to account for the multiply-accumulate operation used in matrix multiplication)

Despite requiring 101x fewer ops and operating over smaller tensors than Case 1, in the example above Case 2 has a wall clock time 34% greater than Case 1.
Are there methods/kernels that can be used which are more optimised towards the later case?

## Environment

Performed using a fresh Google Colab instance.

PyTorch version: 1.5.1+cu101
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.5.1+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.6.1+cu101
[conda] Could not collect

## Additional context
Using `opt_einsum==3.2.1` did not improve performance on Case 2.
```
from opt_einsum import contract
c = torch.randn(10000, 100, 1).cuda()
d = torch.randn(10000, 100, 1, 3).cuda()

%%time
for _ in range(10000):
    contract("bic, bicf->bif", c, d)
    torch.cuda.synchronize()
```
`CPU times: user 39 s, sys: 23.7 s, total: 1min 2s
Wall time: 1min 2s`

cc @ngimel @vincentqb @vishwakftw @jianyuh @nikitaved @pearu @VitalyFedyunin

### LINKED ISSUE #45854: torch.einsum ellipsis broadcasting is too strict and differs from numpy
## 🐛 Bug

torch.einsum requires ellipsis to represent a fixed number of dimensions across all operands while numpy is more flexible for broadcasting.

## To Reproduce

```
>>> np.einsum('i...k,k...j->i...j', torch.rand(2, 1, 3), torch.rand(3, 2)).shape
(2, 1, 2)

>>> torch.einsum('i...k,k...j->i...j', torch.rand(2, 1, 3), torch.rand(3, 2)).shape
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/scratch/heitorschueroff/pytorch/torch/functional.py", line 344, in einsum
    return _VF.einsum(equation, operands)  # type: ignore
RuntimeError: ellipsis must represent 1 dimensions in all terms
```

## Expected behavior

torch.einsum should match numpy.einsum behavior

## Environment

PyTorch version: 1.8.0a0+3ab88c3
Is debug build: True
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
Clang version: Could not collect
CMake version: version 3.14.0

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.105
GPU models and configuration: 
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 418.116.00
cuDNN version: Could not collect
HIP runtime version: N/A

cc @mruberry @rgommers @vincentqb @vishwakftw @jianyuh @nikitaved @pearu

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48184 [CUDA] Optimize sumproduct_pair for certain inputs
* **#47860 Fixed einsum compatibility/performance issues (#46398)**

Summary:
This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option as requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

Partial fix for #32591 with benchmark below. There are still cases (see comments of issue) where it's slower.
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [x] update docs and add more comments
- [x] add tests for error cases

Differential Revision: [D24923679](https://our.internmc.facebook.com/intern/diff/D24923679)

### CODE REVIEW

File: aten/src/ATen/native/Linear.cpp [hunk 1]
```diff
@@ -136,241 +136,333 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
...
+  // Mapping of label to index in the permuted tensors (out_dims + sum_dims)
+  // This will be used for aligning the dimensions of all input operands
+  std::vector<int> label_perm_index(total_labels, -1);
+  
+  // Current index in the permuted shape
+  int perm_index = 0;
+
+  // Start index of ellipsis dimensions in the permuted shape
+  int64_t ell_index = 0;
+  found_ell = false;
```
Comment (heitorschueroff) [inline aten/src/ATen/native/Linear.cpp:258]: My fix starts here, I store whether we found ellipsis for the output.

File: aten/src/ATen/native/Linear.cpp [hunk 2]
```diff
@@ -136,241 +136,333 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
...
-  int64_t position = num_output_dims;
-  for (int64_t i = 0; i < num_total_idxes; i++) {
-    if (idxes_to_preprocessed_dims[i]==-1) {
-      idxes_to_preprocessed_dims[i] = position;
-      position++;
+
+  // Save output size before adding sum dims
+  int out_size = perm_index;
+
+  // If ellipsis is not part of the output, add to contraction dimensions
```
Comment (heitorschueroff) [inline aten/src/ATen/native/Linear.cpp:318]: If we did not have ellipsis for output, I add it to contraction list to be summed out

File: test/test_linalg.py [hunk 3]
```diff
@@ -1324,6 +1324,147 @@ def test_dot_invalid_args(self, device):
...
+        # Test ellipsis broadcasting
+        check('...', 1, expected_output=1)
+        check('...->', 1, expected_output=1)
+        check('...->...', 1, expected_output=1)
+        check('...', [1], expected_output=[1])
+        check('...->', [1], expected_output=1)
+        check('i...->i', [1], expected_output=[1])
+        check('i...->...i', [1], expected_output=[1])
+        check('...a->', [[2], [4]], expected_output=6)
+        check('a...b->ab', [[[1], [2]], [[3], [4]]], expected_output=[[3], [7]])
```
Comment (heitorschueroff) [inline test/test_linalg.py:1437]: The only other changes are in the testing, I'm testing that ellipsis now can be summed out.

**Comment by heitorschueroff** [comment]
@mruberry You can refer to this PR https://github.com/pytorch/pytorch/pull/47852 to see the difference with the previous one.

**Review by mruberry — APPROVED** [review]
Cool!

File: torch/functional.py [hunk 4]
```diff
@@ -262,76 +262,102 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
...
-    tensor([[-1.0864,  0.7292,  0.0569],
-            [-0.9725, -1.0270,  0.6493],
-            [ 0.5832, -1.1716, -1.5084],
-            [ 0.4041, -1.1690,  0.8570]])
-
-    >>> A = torch.randn(2, 3, 4, 5)
-    >>> torch.einsum('...ij->...ji', A).shape # batch permute
-    torch.Size([2, 3, 5, 4])
-"""
+    Sums the product of the elements of the input :attr:`operands` along dimensions specified using a notation
```
Comment (mruberry) [inline torch/functional.py:265]: Make sure to update the docs to reflect this change, too.

**Pull request closed without merging**